File: genetic-helloworld.go
// http://play.golang.org/p/Xg9f0lM40n 

package main

import (
    "fmt"
    "math"
    "math/rand"
    "sort"
    "time"
)

const GA_POPSIZE int = 2048
const GA_MAXITER int = 16384
const GA_ELITRATE float64 = 0.10
const GA_MUTATIONRATE float64 = 0.25

// ideally GA_MUTATION would be an int32, but that requires truncating,
// which we can't do at initialization time
const GA_MUTATION float64 = math.MaxInt32 * GA_MUTATIONRATE
const GA_TARGET string = "Hello world!"

type ga_struct struct {
    str     []byte
    fitness uint
}

type ga_vector []ga_struct

func init_population(population, buffer *ga_vector) {
    tsize := len(GA_TARGET)

    *population = make(ga_vector, GA_POPSIZE)
    for i := 0; i < GA_POPSIZE; i++ {
        citizen := &(*population)[i]

        citizen.fitness = 0

        citizen.str = make([]byte, tsize)
        for j := 0; j < tsize; j++ {
            citizen.str[j] = byte(rand.Intn(90) + 32)
        }
    }
    *buffer = make(ga_vector, GA_POPSIZE)
}

func calc_fitness(population ga_vector) {
    target := GA_TARGET
    tsize := len(target)

    for i := 0; i < GA_POPSIZE; i++ {
        fitness := uint(0)
        for j := 0; j < tsize; j++ {
            fitness += uint(math.Abs(float64(population[i].str[j] - target[j])))
        }
        population[i].fitness = fitness
    }
}

func (s ga_vector) Len() int {
    return len(s)
}

func (s ga_vector) Less(i, j int) bool {
    return s[i].fitness < s[j].fitness
}

func (s ga_vector) Swap(i, j int) {
    s[i], s[j] = s[j], s[i]
}

func sort_by_fitness(population ga_vector) {
    sort.Sort(population)
}

func elitism(population, buffer ga_vector, esize int) {
    copy(buffer[:esize], population[:esize])
}

func mutate(member ga_struct) {
    tsize := len(GA_TARGET)
    ipos := rand.Intn(tsize)
    delta := byte(rand.Intn(90) + 32)

    member.str[ipos] = ((member.str[ipos] + delta) % 122)
}

func mate(population, buffer ga_vector) {
    esize := int(math.Trunc(float64(GA_POPSIZE) * GA_ELITRATE))
    tsize := len(GA_TARGET)

    elitism(population, buffer, esize)

    // Mate the rest
    for i := esize; i < GA_POPSIZE; i++ {
        i1 := rand.Intn(GA_POPSIZE / 2)
        i2 := rand.Intn(GA_POPSIZE / 2)
        spos := rand.Intn(tsize)

        if len(buffer[i].str) < tsize {
            buffer[i].str = make([]byte, tsize)
        }
        copy(buffer[i].str, population[i1].str[:spos])
        copy(buffer[i].str[spos:], population[i2].str[spos:])

        if float64(rand.Int31()) < GA_MUTATION {
            mutate(buffer[i])
        }
    }
}

func print_best(gav ga_vector) {
    fmt.Printf("Best: %s (%d)\n", gav[0].str, gav[0].fitness)
}

func swap(population, buffer *ga_vector) {
    *population, *buffer = *buffer, *population
}

func main() {
    rand.Seed(time.Now().Unix())

    var population, buffer ga_vector
    init_population(&population, &buffer)

    for i := 0; i < GA_MAXITER; i++ {
        calc_fitness(population)
        sort_by_fitness(population)
        print_best(population)

        if population[0].fitness == 0 {
            break
        }

        mate(population, buffer)
        swap(&population, &buffer)
    }
}

File: genetic-helloworld.js
// http://jsfiddle.net/zENKM/

var pop_size = 256; // 2048;
var max_iterations = 512; // 16384;
var elite_rate = 0.10;
var mutation_rate = 0.25;
var mutation = function() {
    return mutation_rate;
}
var target = 'Hello world!';

var struct = function( str, fitness ) {
    return {
        str: str,
        fitness: fitness
    };
}

var init_population = function ( population, buffer ) {
    var size = target.length;

    for ( var i = 0; i < pop_size; i++ ) {
        var citizen = struct('', 0);
        var str = [];

        for ( j = 0; j < size; j++ ) {
            str[j] = String.fromCharCode(Math.floor(Math.random() * 256));

        }

        citizen.str = str.join('');

        population.push(citizen);
        buffer.push(citizen);
    }
}

var calc_fitness = function ( population ) {
    var size = target.length;

    for( var i = 0; i < pop_size; i++ ) {
        var fitness = 0;
        var str = population[i].str;

        for ( j = 0; j < size; j++ ) {
            fitness += Math.abs( str.charCodeAt(j) - target.charCodeAt(j) );
        }

        population[i].fitness = fitness;
    }
}

var fitness_sort = function ( x, y ) {
    return x.fitness - y.fitness;
}

var sort_by_fitness = function ( population ) {
    population.sort(fitness_sort);
}

var elitism = function ( population, buffer, esize )
{
    for ( var i = 0; i < esize; i++ ) {
        buffer[i].str = population[i].str;
        buffer[i].fitness = population[i].fitness;
    }
}

var mutate = function ( member ) {
    var size = target.length;
    var ipos = Math.floor(Math.random() * size);
    var delta = Math.floor(Math.random() * 256);
    var str = member.str.split('');

    str[ipos] = String.fromCharCode((member.str.charCodeAt(ipos) + delta) % 255 );

    member.str = str.join('');
}

var mate = function ( population, buffer ) {
    var esize = pop_size * elite_rate;
    var tsize = target.length;

    elitism ( population, buffer, esize );

    for ( var i = Math.floor(esize); i < pop_size; i++ )
    {
        var i1 = Math.floor(Math.random() * (pop_size / 2));
        var i2 = Math.floor(Math.random() * (pop_size / 2));
        var spos = Math.floor(Math.random() * tsize);

        buffer[i].str = population[i1].str.substr(0, spos) +
            population[i2].str.substr(spos, tsize - spos);

        if ( Math.random() < mutation() ) {
            mutate(buffer[i]);
        }
    }
}

var print_best = function ( population, i ) {
    console.log('Generation ' + i + ' Best: ' + population[0].str + ' (' + population[0].fitness + ')');
}

var swap = function ( population, buffer ) {
    var temp = population;

    population = buffer;
    buffer = temp;
}

var main = function () {
    var population = [];
    var buffer = [];
    var i = 0;

    init_population(population, buffer);

    while ( i < max_iterations ) {

        calc_fitness(population);
        sort_by_fitness(population);
        print_best(population, i);

        if ( population[0].fitness == 0 ) {
            break;
        }

        mate(population, buffer);
        swap(population, buffer);

        i++;
    }
}

main();
